chain/client/src/client.rs

```diff
@@ -662,6 +662,26 @@ impl Client {
             PartialEncodedChunk { header, parts: response.parts, receipts: response.receipts };
         self.process_partial_encoded_chunk(partial_chunk)
     }
+
+    pub fn process_partial_encoded_chunk_forward(
+        &mut self,
+        forward: PartialEncodedChunkForwardMsg,
+    ) -> Result<Vec<AcceptedBlock>, Error> {
+        let header = match self.shards_mgr.get_partial_encoded_chunk_header(&forward.chunk_hash) {
+            Ok(header) => Ok(header),
+            Err(near_chunks::Error::UnknownChunk) => {
+                // We don't know this chunk yet; cache the forwarded part
+                // to be used after we get the header.
+                self.shards_mgr.insert_chunk_forward(forward);
```

Comment:
This doesn't work naively.
With this logic there's an attack vector where we send bogus chunks (with completely invalid hashes) to all the block producers, and invalidate their caches faster than they get to receive the actual chunk header.
The `forward` message needs to contain the signature of the original chunk producer, and some basic info needed to verify the signature is correct (the `height_created` is potentially sufficient -- it should be enough to both recover the epoch id and then the block producer id), and we should not cache anything for which the signature doesn't check out.

We should also only cache one hash per height per shard_id and only for future heights, so that a malicious block producer cannot spam the cache by producing lots of chunks and sending headers for which the signature does check out.